File: NutriSmart/app/src/main/java/com/example/nutrismart/ui/screens/onboarding/OnboardingViewModel.kt
package com.example.nutrismart.ui.screens.onboarding

import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.nutrismart.data.UserSession
import com.example.nutrismart.data.model.ActivityLevel
import com.example.nutrismart.data.model.Currency
import com.example.nutrismart.data.model.DietaryPreference
import com.example.nutrismart.data.model.Gender
import com.example.nutrismart.data.model.MedicalCondition
import com.example.nutrismart.data.model.OnboardingRequest
import com.example.nutrismart.data.repository.UserRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import javax.inject.Inject

@HiltViewModel
class OnboardingViewModel @Inject constructor(
    private val userRepository: UserRepository
) : ViewModel() {

    var birthDate by mutableStateOf(LocalDate.now().minusYears(18))
    var gender by mutableStateOf(Gender.MALE)
    var height by mutableStateOf("")
    var weight by mutableStateOf("")

    var targetWeight by mutableStateOf("")
    var activityLevel by mutableStateOf(ActivityLevel.SEDENTARY)
    var budget by mutableStateOf("")

    var isImperial by mutableStateOf(false)
    var currency by mutableStateOf(Currency.RON)

    var selectedDietaryPreferences = mutableStateListOf<DietaryPreference>()
    var selectedMedicalConditions = mutableStateListOf<MedicalCondition>()
    var selectedDislikedFoods by mutableStateOf<Set<String>>(emptySet())

    var foodSearchQuery by mutableStateOf("")
    var isSearchingFoods by mutableStateOf(false)
    var foodSuggestions = mutableStateListOf<String>()

    var isLoading by mutableStateOf(false)
    var errorMessage by mutableStateOf<String?>(null)
    var isComplete by mutableStateOf(false)
    var loadingMessage by mutableStateOf("Saving profile...")

    private val token: String get() = UserSession.token ?: ""
    private val userId: Long get() = UserSession.currentUserId

    fun loadUserData() {
        if (userId == -1L) return

        isLoading = true
        loadingMessage = "Loading your details..."

        viewModelScope.launch {
            try {
                val response = userRepository.getUser(token, userId)

                if (response.isSuccessful) {
                    response.body()?.let { user ->
                        user.dateOfBirth?.let {
                            runCatching { birthDate = LocalDate.parse(it) }
                        }
                        user.gender?.let { gender = it }
                        user.height?.let { height = it.toString() }
                        user.weight?.let { weight = it.toString() }
                        user.targetWeight?.let { targetWeight = it.toString() }
                        user.activityLevel?.let { activityLevel = it }
                        user.maxDailyBudget?.let { budget = it.toString() }
                        user.currency?.let { currency = it }
                        isImperial = user.isImperial

                        user.dietaryPreferences?.let {
                            selectedDietaryPreferences.clear()
                            selectedDietaryPreferences.addAll(it)
                        }
                        user.medicalConditions?.let {
                            selectedMedicalConditions.clear()
                            selectedMedicalConditions.addAll(it)
                        }
                        user.dislikedFoods?.let {
                            selectedDislikedFoods = it.toSet()
                        }
                    }
                }
            } catch (e: Exception) {
                Log.e("NutriSmart", "Error loading user data", e)
            } finally {
                isLoading = false
            }
        }
    }

    private val messages = listOf(
        "Analyzing your profile...",
        "Calculating target calories...",
        "Generating 14-day AI recipes...",
        "Finalizing your plan...",
        "Do not panic if it looks stuck\nAI is working"
    )

    fun toggleDiet(diet: DietaryPreference) {
        if (selectedDietaryPreferences.contains(diet)) {
            selectedDietaryPreferences.remove(diet)
        } else {
            selectedDietaryPreferences.add(diet)
        }
    }

    fun toggleCondition(condition: MedicalCondition) {
        if (selectedMedicalConditions.contains(condition)) {
            selectedMedicalConditions.remove(condition)
        } else {
            selectedMedicalConditions.add(condition)
        }
    }

    fun toggleDislikedFood(food: String) {
        val currentSet = selectedDislikedFoods.toMutableSet()
        if (currentSet.contains(food)) {
            currentSet.remove(food)
        } else {
            currentSet.add(food)
        }
        selectedDislikedFoods = currentSet
    }

    fun onFoodSearchQueryChanged(newQuery: String) {
        foodSearchQuery = newQuery
        if (newQuery.isBlank()) {
            foodSuggestions.clear()
            return
        }

        searchFoods(newQuery)
    }

    private var searchJob: Job? = null
    private fun searchFoods(query: String) {
        searchJob?.cancel()
        searchJob = viewModelScope.launch {
            delay(400)
            isSearchingFoods = true
            try {
                Log.d("NutriSmart", "Searching for: $query")
                val response = userRepository.searchFoods(token, query)
                
                if (response.isSuccessful) {
                    val data = response.body()
                    Log.d("NutriSmart", "Found: ${data?.size} items")
                    foodSuggestions.clear()
                    data?.let { foodSuggestions.addAll(it) }
                } else {
                    Log.e("NutriSmart", "Search failed: ${response.code()}")
                }
            } catch (e: Exception) {
                Log.e("NutriSmart", "Search error", e)
            } finally {
                isSearchingFoods = false
            }
        }
    }

    fun submitProfile() {
        val hValue = height.replace(',', '.').toDoubleOrNull()
        val wValue = weight.replace(',', '.').toDoubleOrNull()
        val twValue = targetWeight.replace(',', '.').toDoubleOrNull()
        val bValue = budget.replace(',', '.').toDoubleOrNull()

        if (hValue == null || wValue == null || twValue == null || bValue == null) {
            errorMessage = "Please fill in all numeric fields correctly."
            return
        }

        if (birthDate.isAfter(LocalDate.now())) {
            errorMessage = "Birth date cannot be in the future."
            return
        }
        val age = java.time.Period.between(birthDate, LocalDate.now()).years
        if (age < 14) {
            errorMessage = "You must be at least 14 years old to use this app."
            return
        }

        val minHeight = if (isImperial) 3.3 else 100.0
        val maxHeight = if (isImperial) 9.8 else 300.0
        if (hValue !in minHeight..maxHeight) {
            errorMessage = "Height must be between $minHeight and $maxHeight ${if (isImperial) "ft" else "cm"}."
            return
        }

        val minWeight = if (isImperial) 33.0 else 15.0
        val maxWeight = if (isImperial) 661.0 else 300.0
        if (wValue !in minWeight..maxWeight) {
            errorMessage = "Weight must be between $minWeight and $maxWeight ${if (isImperial) "lbs" else "kg"}."
            return
        }
        if (twValue !in minWeight..maxWeight) {
            errorMessage = "Target weight must be between $minWeight and $maxWeight ${if (isImperial) "lbs" else "kg"}."
            return
        }

        val minBudget = if (currency == Currency.RON) 20.0 else 10.0
        if (bValue < minBudget) {
            errorMessage = "Minimum daily budget is $minBudget ${currency.name}."
            return
        }

        isLoading = true
        errorMessage = null
        loadingMessage = "Saving profile..."

        viewModelScope.launch {
            try {
                val dateString = birthDate.format(DateTimeFormatter.ISO_LOCAL_DATE)

                val request = OnboardingRequest(
                    dateOfBirth = dateString,
                    gender = gender,
                    height = hValue,
                    weight = wValue,
                    targetWeight = twValue,
                    activityLevel = activityLevel,
                    maxDailyBudget = bValue,
                    dietaryPreferences = selectedDietaryPreferences.toList(),
                    medicalConditions = selectedMedicalConditions.toList(),
                    dislikedFoods = selectedDislikedFoods.toList(),
                    isImperial = isImperial,
                    currency = currency
                )

                val response = userRepository.completeProfile(token, request)

                if (response.isSuccessful) {
                    if (userId == -1L) {
                        errorMessage = "Error: User ID not found in session."
                        isLoading = false
                        return@launch
                    }
                    startPlanGeneration(userId)
                } else {
                    val errorBody = response.errorBody()?.string()
                    errorMessage = if (!errorBody.isNullOrEmpty() && errorBody.contains("message")) {
                        "Error ${response.code()}: AI profile save failed."
                    } else {
                        "Error: ${response.code()} - ${response.message()}"
                    }
                    isLoading = false
                }

            } catch (e: Exception) {
                errorMessage = "Connection error: ${e.message}"
                isLoading = false
            }
        }
    }

    private fun startPlanGeneration(userId: Long) {
        viewModelScope.launch {
            try {
                val startResponse = userRepository.startPlanGeneration(token, userId)

                if (startResponse.isSuccessful) {
                    val messageJob = launchMessageRotation()
                    pollForStatus(userId, messageJob)
                } else {
                    errorMessage = "Failed to start AI generation."
                    isLoading = false
                }
            } catch (e: Exception) {
                errorMessage = "Network error: ${e.message}"
                isLoading = false
            }
        }
    }

    private fun launchMessageRotation(): Job {
        return viewModelScope.launch {
            var index = 0
            while (isActive) {
                loadingMessage = messages[index]
                if (index < messages.size - 1) index++
                delay(6000)
            }
        }
    }

    private suspend fun pollForStatus(userId: Long, messageJob: Job) {
        var isDone = false
        while (!isDone && viewModelScope.isActive) {
            delay(5000)
            try {
                val statusResponse = userRepository.checkGenerationStatus(token, userId)

                if (statusResponse.isSuccessful) {
                    val status = statusResponse.body()?.get("status") ?: "UNKNOWN"
                    when (status) {
                        "COMPLETED" -> {
                            isDone = true
                            messageJob.cancel()
                            loadingMessage = "Plan generated successfully!"
                            delay(1000)
                            isLoading = false
                            isComplete = true
                        }
                        "FAILED" -> {
                            isDone = true
                            messageJob.cancel()
                            errorMessage = "AI generation failed. Please try again."
                            isLoading = false
                        }
                    }
                }
            } catch (e: Exception) {
                println(e)
            }
        }
    }
}

File: NutriSmartIOS/ui/screens/onboarding/OnboardingViewModel.swift
//
//  OnboardingViewModel.swift
//  NutriSmartIOS
//

import Foundation
import SwiftUI
import Combine

@Observable
class OnboardingViewModel {
    var birthDate = Calendar.current.date(byAdding: .year, value: -18, to: Date()) ?? Date()
    var gender: Gender = .MALE
    var height = ""
    var weight = ""
    var targetWeight = ""
    var activityLevel: ActivityLevel = .SEDENTARY
    var budget = ""
    var isImperial = false
    var currency: Currency = .RON
    var selectedDietaryPreferences: [DietaryPreference] = []
    var selectedMedicalConditions: [MedicalCondition] = []
    var selectedDislikedFoods: Set<String> = []
    var foodSearchQuery = ""
    var isSearchingFoods = false
    var foodSuggestions: [String] = []
    var isLoading = false
    var errorMessage: String? = nil
    var isComplete = false
    var loadingMessage = "Saving profile..."
    
    private let userRepository = UserRepository.shared
    private let token: String
    private let userId: Int
    private var searchTask: Task<Void, Never>? = nil
    private var messageTask: Task<Void, Never>? = nil
    
    private let messages = [
        "Analyzing your profile...", "Calculating target calories...",
        "Generating 14-day AI recipes...", "Finalizing your plan...",
        "Do not panic if it looks stuck\nAI is working"
    ]
    
    init() {
        self.token = SessionManager.shared.fetchAuthToken() ?? ""
        self.userId = SessionManager.shared.fetchUserId()
    }
    
    func loadUserData() {
        guard userId != -1 else { return }
        isLoading = true
        loadingMessage = "Loading your details..."
        userRepository.getUser(userId: userId) { [weak self] result in
            DispatchQueue.main.async {
                self?.isLoading = false
                if case .success(let user) = result {
                    if let dobStr = user.dateOfBirth {
                        let formatter = ISO8601DateFormatter()
                        formatter.formatOptions = [.withFullDate, .withDashSeparatorInDate]
                        if let date = formatter.date(from: dobStr) { self?.birthDate = date }
                    }
                    if let gender = user.gender { self?.gender = gender }
                    if let h = user.height { self?.height = String(h) }
                    if let w = user.weight { self?.weight = String(w) }
                    if let tw = user.targetWeight { self?.targetWeight = String(tw) }
                    if let al = user.activityLevel { self?.activityLevel = al }
                    if let b = user.maxDailyBudget { self?.budget = String(b) }
                    if let c = user.currency { self?.currency = c }
                    self?.isImperial = user.isImperial
                    self?.selectedDietaryPreferences = user.dietaryPreferences ?? []
                    self?.selectedMedicalConditions = user.medicalConditions ?? []
                    self?.selectedDislikedFoods = Set(user.dislikedFoods ?? [])
                }
            }
        }
    }
    
    func toggleDiet(_ diet: DietaryPreference) {
        if let index = selectedDietaryPreferences.firstIndex(of: diet) { selectedDietaryPreferences.remove(at: index) }
        else { selectedDietaryPreferences.append(diet) }
    }
    func toggleCondition(_ condition: MedicalCondition) {
        if let index = selectedMedicalConditions.firstIndex(of: condition) { selectedMedicalConditions.remove(at: index) }
        else { selectedMedicalConditions.append(condition) }
    }
    func toggleDislikedFood(_ food: String) {
        if selectedDislikedFoods.contains(food) { selectedDislikedFoods.remove(food) }
        else { selectedDislikedFoods.insert(food) }
    }
    func onFoodSearchQueryChanged(_ newQuery: String) {
        foodSearchQuery = newQuery
        if newQuery.isEmpty { foodSuggestions = []; return }
        searchFoods(query: newQuery)
    }
    private func searchFoods(query: String) {
        searchTask?.cancel()
        searchTask = Task {
            try? await Task.sleep(nanoseconds: 400 * 1_000_000)
            if Task.isCancelled { return }
            await MainActor.run { isSearchingFoods = true }
            userRepository.searchFoods(query: query) { [weak self] result in
                DispatchQueue.main.async {
                    self?.isSearchingFoods = false
                    if case .success(let suggestions) = result { self?.foodSuggestions = suggestions }
                }
            }
        }
    }
    func submitProfile() {
        guard let hValue = Double(height.replacingOccurrences(of: ",", with: ".")),
              let wValue = Double(weight.replacingOccurrences(of: ",", with: ".")),
              let twValue = Double(targetWeight.replacingOccurrences(of: ",", with: ".")),
              let bValue = Double(budget.replacingOccurrences(of: ",", with: ".")) else {
            errorMessage = "Please fill in all numeric fields correctly."
            return
        }
        if birthDate > Date() { errorMessage = "Birth date cannot be in the future."; return }
        isLoading = true; errorMessage = nil; loadingMessage = "Saving profile..."
        let formatter = DateFormatter(); formatter.dateFormat = "yyyy-MM-dd"
        let request = OnboardingRequest(
            dateOfBirth: formatter.string(from: birthDate), gender: gender, height: hValue, weight: wValue,
            targetWeight: twValue, activityLevel: activityLevel, maxDailyBudget: bValue,
            dietaryPreferences: selectedDietaryPreferences, medicalConditions: selectedMedicalConditions,
            dislikedFoods: Array(selectedDislikedFoods), isImperial: isImperial, currency: currency
        )
        userRepository.completeProfile(request: request) { [weak self] result in
            DispatchQueue.main.async {
                switch result {
                case .success(_): self?.startPlanGeneration()
                case .failure(let error): self?.isLoading = false; self?.errorMessage = "Error saving profile: \(error)"
                }
            }
        }
    }
    private func startPlanGeneration() {
        isLoading = true; loadingMessage = messages[0]
        launchMessageRotation()
        pollForStatus()
    }
    private func launchMessageRotation() {
        messageTask?.cancel()
        messageTask = Task {
            var index = 0
            while !Task.isCancelled {
                await MainActor.run {
                    loadingMessage = messages[index]
                    if index < messages.count - 1 { index += 1 }
                }
                try? await Task.sleep(nanoseconds: 6 * 1_000_000_000)
            }
        }
    }
    private func pollForStatus() {
        Task {
            var isDone = false
            while !isDone && !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 5 * 1_000_000_000)
                ApiClient.shared.request(endpoint: "nutrition/status/\(userId)") { [weak self] (result: Result<[String: String], NetworkError>) in
                    DispatchQueue.main.async {
                        if case .success(let dict) = result {
                            let status = dict["status"] ?? "UNKNOWN"
                            if status == "COMPLETED" {
                                isDone = true; self?.messageTask?.cancel()
                                self?.loadingMessage = "Plan generated successfully!"; self?.isLoading = false; self?.isComplete = true
                            } else if status == "FAILED" {
                                isDone = true; self?.messageTask?.cancel()
                                self?.errorMessage = "AI generation failed. Please try again."; self?.isLoading = false
                            }
                        }
                    }
                }
            }
        }
    }
}
